### ISSUE TITLE
Code coverage aggregation

### ISSUE THREAD

**Issue by daviddoria** [bilke/cmake-modules#8]
Is there a way to aggregate the results of the coverage analysis? That is, it looks like CodeCoverage.cmake is a per-target (one per add_test()?) function. Say I have 10 subdirectories each with their own set of add_test() calls - how do I see the result as a coverage percentage for the whole project?

**Comment by bilke** [comment]
Yes, the coverage analysis is done per-target. What I do is to create a custom target which runs `ctest`:

```cmake
add_custom_target(ctest COMMAND ${CMAKE_CTEST_COMMAND})
```

Then I set this up as a coverage target:

```cmake
SETUP_TARGET_FOR_COVERAGE_COBERTURA(ctest_coverage_cobertura ctest "ctest_coverage_cobertura_results" "-j;${PROCESSOR_COUNT}")
```

This will produce an aggregated coverage report of everything which is part of the `ctest`-run.

Additionally I setup more coverage targets (e.g. GoogleTest runner) which produce each their own coverage report and at the end I aggregate them in Jenkins with the [Cobertura Code Coverage Publisher](https://wiki.jenkins-ci.org/display/JENKINS/Cobertura+Plugin).

**Comment by daviddoria** [comment]
Thanks, I'll try this today. I guess I'm surprised that the coverage analyzer is able to gather the output of multiple executables just because they are launched by the same executable.

Did I understand correctly that you use both CTest and GTest in the same project? How do you decide what to do in each testing system?

**Comment by bilke** [comment]
We use GTest for unit tests and CTest for end-to-end (executable) testing.

**Comment by daviddoria** [comment]
Hm, I must be doing something wrong. Here is my CMakeLists.txt:

Project(CoverageExampleProject)
include(CTest)

set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_MODULE_PATH})
INCLUDE(CodeCoverage)

SET(CMAKE_CXX_FLAGS="-g -O0 --coverage")
SET(CMAKE_C_FLAGS="-g -O0 --coverage")
SET(CMAKE_EXE_LINKER_FLAGS="--coverage")
 
add_executable(CoverageExample CoverageExample.cpp)
target_link_libraries(CoverageExample gcov)

add_test(CoverageExampleTest CoverageExample)

add_custom_target(ctestTarget COMMAND ${CMAKE_CTEST_COMMAND})

SETUP_TARGET_FOR_COVERAGE(coverageTarget ctestTarget "testResults")

and the output:

~/build/Examples/c++/CoverageExample2 $ make coverageTarget
-- Configuring done
-- Generating done
-- Build files have been written to: /home/doria/build/Examples/c++/CoverageExample2
[100%] Resetting code coverage counters to zero.
Processing code coverage counters and generating report.
Deleting all .da files in . and subdirectories
Done.
make[3]: ctestTarget: Command not found
make[3]: *** [CMakeFiles/coverageTarget] Error 127
make[2]: *** [CMakeFiles/coverageTarget.dir/all] Error 2
make[1]: *** [CMakeFiles/coverageTarget.dir/rule] Error 2
make: *** [coverageTarget] Error 2

Note that 'make ctestTarget' seems to work fine (it runs ctest). Any suggestions?

**Comment by bilke** [comment]
Ah thanks, this is a bug:

The script assumes that the second parameter is a valid target as well as a binary located in `CMAKE_BINARY_DIR`:

```cmake
SETUP_TARGET_FOR_COVERAGE(coverageTarget ctestTarget "testResults")
```

Obviously `ctestTarget` is a target but not a binary. I have not experienced that because my target is simply called `ctest` which also is not a binary created from the project but a global binary from CMake. I will think about a solution tomorrow ...

**Comment by daviddoria** [comment]
Makes sense. Let me know if you'd like me to test a patch.

For the time being, I changed to :

add_custom_target(ctest COMMAND ${CMAKE_CTEST_COMMAND})
SETUP_TARGET_FOR_COVERAGE(coverageTarget ctest "testResults")

but I get:

~/build/Examples/c++/CoverageExample2 $ make coverageTarget
[100%] Resetting code coverage counters to zero.
Processing code coverage counters and generating report.
Deleting all .da files in . and subdirectories
Done.
Test project /home/doria/build/Examples/c++/CoverageExample2
    Start 1: CoverageExampleTest
1/1 Test #1: CoverageExampleTest ..............   Passed    0.00 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.00 sec
Capturing coverage data from .
Found gcov version: 4.8.4
Scanning . for .gcda files ...
geninfo: ERROR: no .gcda files found in .!
make[3]: *** [CMakeFiles/coverageTarget] Error 255
make[2]: *** [CMakeFiles/coverageTarget.dir/all] Error 2
make[1]: *** [CMakeFiles/coverageTarget.dir/rule] Error 2
make: *** [coverageTarget] Error 2

Any thoughts?

**Comment by bilke** [comment]
Maybe you are missing `add_definitions(-fprofile-arcs -ftest-coverage)`?

**Comment by bilke** [comment]
I have updated the coverage scripts in another project. I will test it there first.

- [CodeCoverage.cmake](https://github.com/bilke/ogs/blob/coverage-fixes/scripts/cmake/cmake/CodeCoverage.cmake)
-[Invocation of CodeCoverage](https://github.com/bilke/ogs/blob/coverage-fixes/scripts/cmake/Coverage.cmake)

**Comment by daviddoria** [comment]
I tried it and noticed two things (I know you haven't tested yet, but I thought it may help):

1) I had to add
include(CMakeParseArguments)
or I got "Unknown CMake command "cmake_parse_arguments"

2) At this line:
add_custom_command(TARGET ${Coverage_NAME} POST_BUILD ....

Coverage_NAME seems to be empty

**Comment by daviddoria** [comment]
(Sorry we have parallel threads going)
Adding
add_definitions(-fprofile-arcs -ftest-coverage)
indeed fixed it. I had:
SET(CMAKE_CXX_FLAGS="-g -O0 -fprofile-arcs -ftest-coverage")
SET(CMAKE_C_FLAGS="-g -O0 -fprofile-arcs -ftest-coverage")
SET(CMAKE_EXE_LINKER_FLAGS="-fprofile-arcs -ftest-coverage")

which I thought was the same thing, but apparently it is not? haha

Also, FYI I learned that
add_definitions(--coverage)
is apparently the more modern alias for the same thing.

David

**Comment by dustingooding** [comment]
I'd like to have aggregate code coverage reporting as well, but I'm unable to use ctest.  I'm using [catkin](https://catkin-tools.readthedocs.io/en/latest/) to build code in a [ROS](http://www.ros.org/) ecosystem and switching to a different build system is prohibitive.  Simply creating a custom target that depends on the multiple `SETUP_TARGET_FOR_COVERAGE` targets results in errors (cannot read various symbols), which I believe is due to the multiple calls to delete/zero in a shared build space.  I think a reasonable approach would be to have three "steps", one that zeros, one that runs the binaries, and one that generates the aggregated report.  The trick will be to get all these targets/commands to properly depend on each other such that you don't have a lot of fixturing in your projects' CMakeLists.txt.

Have you made any progress on this aggregation idea?  If I get it to work, would you like a PR from me?

**Comment by CiaranWelsh** [comment]
Is there a complete example somewhere of how to do this using `googletest`? Thanks.

**Comment by rayment** [comment]
Hi, I've written a small tutorial in #88 with a basic template to get started with `ctest` and `doctest`, though `doctest` is entirely optional and you can use a different test framework should you desire.

Aggregation with `ctest` actually works well and I have had no issues so far.
